modular_skyrat/modules/modular_implants/code/soulcatcher/handheld_soulcatcher.dm

```diff
@@ -0,0 +1,128 @@
+	icon = 'modular_skyrat/modules/modular_implants/icons/obj/devices.dmi'
+	icon_state = "soulcatcher-device"
+	inhand_icon_state = "electronic"
+	lefthand_file = 'icons/mob/inhands/items/devices_lefthand.dmi'
+	righthand_file = 'icons/mob/inhands/items/devices_righthand.dmi'
+	/// What soulcatcher datum is associated with this item?
+	var/datum/component/soulcatcher/linked_soulcatcher
+
+/obj/item/handheld_soulcatcher/attack_self(mob/user, modifiers)
+	linked_soulcatcher.ui_interact(user)
+
+/obj/item/handheld_soulcatcher/New(loc, ...)
+	. = ..()
+	linked_soulcatcher = AddComponent(/datum/component/soulcatcher)
+
+/obj/item/handheld_soulcatcher/Destroy(force)
+	if(linked_soulcatcher)
+		qdel(linked_soulcatcher)
+
+	return ..()
+
+/obj/item/handheld_soulcatcher/attack(mob/living/target_mob, mob/living/user, params)
+	if(!target_mob)
+		return ..()
+
+	var/datum/component/previous_body/body_component = target_mob.GetComponent(/datum/component/previous_body)
+	if(body_component)
+		linked_soulcatcher.scan_body(body_component, user)
+		return TRUE
+
+	if(!target_mob.mind || !target_mob.ckey)
+		to_chat(user, span_warning("You are unable to remove a mind from an empty body."))
+		return FALSE
+
+	if(target_mob.stat == DEAD) //We can temporarily store souls of dead mobs.
+		target_mob.ghostize(TRUE) //Incase they are staying in the body.
+		var/mob/dead/observer/target_ghost = target_mob.get_ghost(TRUE, TRUE)
+		if(!target_ghost)
+			to_chat(user, span_warning("You are unable to get the soul of [target_mob]!"))
+			return FALSE
+
+		var/datum/soulcatcher_room/target_room = tgui_input_list(user, "Choose a room to send [target_mob]'s soul to.", name, linked_soulcatcher.soulcatcher_rooms, timeout = 30 SECONDS)
+		if(!target_room)
+			return FALSE
+
+		if(tgui_alert(target_ghost, "[user] wants to transfer you to [target_room] inside of a soulcatcher, do you accept?", name, list("Yes", "No"), 30 SECONDS, autofocus = TRUE) != "Yes")
+			to_chat(user, span_warning("[target_mob] doesn't seem to want to enter."))
+			return FALSE
+
+		if(!target_room.add_soul_from_ghost(target_ghost))
+			return FALSE
+
+		body_component = target_mob.GetComponent(/datum/component/previous_body)
+		if(!body_component)
+			return FALSE
+
+		var/turf/source_turf = get_turf(user)
+		log_admin("[key_name(user)] used [src] to put [key_name(target_mob)]'s mind into a soulcatcher at [AREACOORD(source_turf)]")
+		linked_soulcatcher.scan_body(body_component, user)
+		return TRUE
+
+	var/datum/soulcatcher_room/target_room = tgui_input_list(user, "Chose a room to send [target_mob]'s soul to.", name, linked_soulcatcher.soulcatcher_rooms, timeout = 30 SECONDS)
```

Comment:
adding a timeout is only a bandaid that SOMETIMES fixes the issue but is easily worked around, you might want to look into custom UI logic that would close/deny input to the window if the target or the user get too far away from the soulcatcher/either of them enter a invalid state (death)

Reply:
That same issue would arguably apply to other TGUI based prompts. I feel that a timeout is adequate enough to prevent most use from bad actors.